File: core/src/main/kotlin/com/example/toyracers/track/Checkpoint.kt
package com.example.toyracers.track

/**
 * An ordered gate across the road.
 *
 * The forward vector records the valid crossing direction for the lap rules.
 */
data class Checkpoint(
    val order: Int,
    val gate: TrackSegment,
    val forwardX: Float,
    val forwardY: Float,
) {
    init {
        require(order >= 0) { "Checkpoint order must not be negative" }
        require(forwardX != 0f || forwardY != 0f) {
            "Checkpoint forward direction must not be zero"
        }
    }
}

File: dart/lib/simulation/track/checkpoint.dart
import '../math/float32.dart';
import 'track_geometry.dart';

/// An ordered gate that must be crossed in its forward direction.
final class Checkpoint {
  Checkpoint({
    required this.order,
    required this.gate,
    required double forwardX,
    required double forwardY,
  }) : forwardX = Float32.narrow(forwardX),
       forwardY = Float32.narrow(forwardY) {
    if (order < 0) {
      throw ArgumentError.value(order, 'order', 'must not be negative');
    }
    if (this.forwardX == 0 && this.forwardY == 0) {
      throw ArgumentError('Checkpoint forward direction must not be zero.');
    }
  }

  final int order;
  final TrackSegment gate;
  final double forwardX;
  final double forwardY;

  @override
  bool operator ==(Object other) =>
      other is Checkpoint &&
      order == other.order &&
      gate == other.gate &&
      forwardX == other.forwardX &&
      forwardY == other.forwardY;

  @override
  int get hashCode => Object.hash(order, gate, forwardX, forwardY);
}
